[raas][torch][jit] Allow not storing the optimized graph

behrang79 opened the pull request:
Summary:
GraphFunction internally stores the optimized graph after generating it and then it is passed into the executor which makes a copy of it. So we store the optimized graph effectively twice.

This diff allows to set a flag to not store the optimized graph inside the GraphFunction.

The code is NoP right now until the flag is enabled.

Test Plan:
I ran SL with this on raas with good memory saving on raas server. From command line:

exmaple model run
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362

I1207 11:04:58.657143 3556226 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 255646 Kb
```

then with flag enabled:
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362 --torch_jit_do_not_store_optimized_graph=true
I1207 11:06:25.245779 3577383 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 165167 Kb
```
So collective with this flag and the flag from D51950418
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362 --torch_jit_do_not_store_optimized_graph=true --torch_jit_enable_profiling_graph_executor=false

I1207 11:09:17.502743 3592345 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 114848 Kb
```

Differential Revision: D51931895

behrang79 commented:
easycla

malfet commented:
/easycla

malfet commented on torch/csrc/jit/api/function_impl.cpp:
```diff
@@ -86,6 +92,28 @@ const c10::FunctionSchema& GraphFunction::getSchema() const {
 
+ std::shared_ptr<Graph> GraphFunction::optimized_graph() const {
+    std::lock_guard<std::recursive_mutex> lock(compile_mutex);
+    std::shared_ptr<Graph> optimized_graph;
+    std::shared_ptr<Graph>* optimized_graph_ptr = nullptr;
+
+    if (FLAGS_torch_jit_do_not_store_optimized_graph) {
+        optimized_graph  = graph_->copy();
+        optimized_graph_ptr = &optimized_graph;
+    } else {
+      auto& optimized_graph_opt = optimized_graphs_[currentSpecialization()];
+      if (optimized_graph_opt) {
+        return *optimized_graph_opt;
+      }
+      optimized_graph_opt = graph_->copy();
+      optimized_graph_ptr = &optimized_graph_opt.value();
+    }
+    if (getGraphExecutorOptimize()) {
+        preoptimizeGraph(*optimized_graph_ptr, force_no_amp_);
+    }
+    return *optimized_graph_ptr;
+ }
```
This makes code a bit shorter (and avoids a rather unusual raw pointer to smart pointer game)
```suggestion
    std::shared_ptr<Graph> optimized_graph;
    std::shared_ptr<Graph>& optimized_graph_ref = &optimized_graph;
    
    if (!FLAGS_torch_jit_do_not_store_optimized_graph) {
      optimized_graph_ref = optimized_graphs_[currentSpecialization()];
       if (optimized_graph_ref) {
         return *optimized_graph_ref;
       }
    } 

    optimized_graph_ref = graph_->copy();

        if (getGraphExecutorOptimize()) {
        preoptimizeGraph(*optimized_graph_ref, force_no_amp_);
    }
    
    return *optimized_graph_ref;
 }
```

malfet commented on torch/csrc/jit/api/function_impl.cpp:
```diff
@@ -86,6 +92,28 @@ const c10::FunctionSchema& GraphFunction::getSchema() const {
 
+std::shared_ptr<Graph> GraphFunction::optimized_graph() const {
+  std::lock_guard<std::recursive_mutex> lock(compile_mutex);
+  std::shared_ptr<Graph> optimized_graph;
+  std::shared_ptr<Graph>* optimized_graph_ptr = nullptr;
+
+  if (FLAGS_torch_jit_do_not_store_optimized_graph) {
+    optimized_graph = graph_->copy();
+    optimized_graph_ptr = &optimized_graph;
+  } else {
+    auto& optimized_graph_opt = optimized_graphs_[currentSpecialization()];
+    if (optimized_graph_opt) {
+      return *optimized_graph_opt;
+    }
+    optimized_graph_opt = graph_->copy();
+    optimized_graph_ptr = &optimized_graph_opt.value();
+  }
+  if (getGraphExecutorOptimize()) {
+    preoptimizeGraph(*optimized_graph_ptr, force_no_amp_);
+  }
+  return *optimized_graph_ptr;
```
```suggestion
  std::shared_ptr<Graph> optimized_graph;

  if (FLAGS_torch_jit_do_not_store_optimized_graph) {
    optimized_graph = graph_->copy();
  } else {
    auto& optimized_graph_opt = optimized_graphs_[currentSpecialization()];
    if (optimized_graph_opt) {
      return *optimized_graph_opt;
    }
    optimized_graph_opt = graph_->copy();
    optionmized_graph = optimized_graph.value();
  }
  if (getGraphExecutorOptimize()) {
    preoptimizeGraph(optimized_graph, force_no_amp_);
  }
  return optimized_graph;
```

behrang79 committed:
[raas][torch][jit] Allow not storing the optimized graph (#115381)

Summary:

GraphFunction internally stores the optimized graph after generating it and then it is passed into the executor which makes a copy of it. So we store the optimized graph effectively twice. 

This diff allows to set a flag to not store the optimized graph inside the GraphFunction.

The code is NoP right now until the flag is enabled.

Test Plan:
I ran SL with this on raas with good memory saving on raas server. From command line:

exmaple model run
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362

I1207 11:04:58.657143 3556226 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 255646 Kb
```

then with flag enabled:
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362 --torch_jit_do_not_store_optimized_graph=true
I1207 11:06:25.245779 3577383 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 165167 Kb
```
So collective with this flag and the flag from D51950418
```
buck run mode/opt-clang  sigrid/predictor/client/localnet:run_model -- --model_id_to_load=953556500 --model_snapshot_to_load=362 --torch_jit_do_not_store_optimized_graph=true --torch_jit_enable_profiling_graph_executor=false

I1207 11:09:17.502743 3592345 SigridPredictorLocalModelFactory.cpp:32] Memory usage for 953556500_362 is 114848 Kb
```

Differential Revision: D51931895
